### ISSUE TITLE
SDL2: solarwolf game should work with SDL2.

### ISSUE THREAD

**Issue by illume** [pygame/pygame#721]
One old game that can be used for compatibility testing is solarwolf.
https://github.com/pygame/solarwolf

**Comment by robertpfeiffer** [comment]
Solarwolf and the aliens example (#717) are both using indexed 8-bit images. Solarwolf has some indexed pngs with per-pixel alpha, i.e. palette colours are not RGB but RGBA. Additionally, solarwolf also uses per-sprite set_alpha() and set_colorkey(). If you convert the pngs to 32bit RGBA with imagemagick, the glitches in the main menu disappear, so it must be related to 8bit surfaces.

I think I have isolated the problem:
```python
import pygame
pygame.init()

screen=pygame.display.set_mode((200,150))
surf=pygame.image.load("solarwolf/data/menu_creds_off.png")
print(surf.get_colorkey())
print(surf.convert().get_colorkey())
```
output with old pygame:
```
pygame 1.9.4
Hello from the pygame community. https://www.pygame.org/contribute.html
(219, 10, 8, 255)
(219, 10, 8, 255)
```
output with pygame+SDL2
```
pygame 1.9.5.dev0
Hello from the pygame community. https://www.pygame.org/contribute.html
(219, 10, 8, 255)
(0, 0, 0, 0)
```

**Comment by robertpfeiffer** [comment]
saving and re-applying the colorkey fixes the issue *and I have no idea why*
```
def my_convert():
    c=img.get_colorkey()
    img=img.convert()
    img.set_colorkey(c)
    return img
```

surf_convert surface.c already calls SDL_SetColorKey.

**Comment by illume** [comment]
Thanks heaps for narrowing this down and coming up with the test case :)

**Comment by robertpfeiffer** [comment]
so this probably means if RLEACCEL is set, then convert still breaks colourkey somehow.

**Comment by robertpfeiffer** [comment]
minimal test case, again

```
import pygame
pygame.init()
screen=pygame.display.set_mode((200,150))
surf=pygame.image.load("data/menu_creds_off.png")

print(surf.get_colorkey())
print(surf.convert().get_colorkey())

surf.set_colorkey(surf.get_colorkey(), pygame.RLEACCEL)
print(surf.get_colorkey())
print(surf.convert().get_colorkey())
```
on pygame+SDL2 commit c474aa9
```
(219, 10, 8, 255)
(219, 10, 8, 255)
(75, 75, 75, 255)
(75, 75, 75, 255)
```
on pygame 1.9.4
```
(219, 10, 8, 255)
(219, 10, 8, 255)
(219, 10, 8, 255)
(219, 10, 8, 255)
```

**Comment by dlon** [comment]
Same result without pygame.RLEACCEL.

**Comment by dlon** [comment]
Another test:

```python3
import pygame
pygame.init()
screen=pygame.display.set_mode((200,150))
surf=pygame.image.load("solarwolf/solarwolf/data/menu_creds_off.png")

print(surf.get_colorkey())
print(surf.map_rgb(surf.get_colorkey()))
print(surf.unmap_rgb(surf.map_rgb(surf.get_colorkey())))
```

SDL2 output:
```
(219, 10, 8, 255)
3
(75, 75, 75, 255)
```

SDL1 output:
```
(219, 10, 8, 255)
0
(219, 10, 8, 255)
```

**Comment by dlon** [comment]
solarwolf seems to be fixed now due to #737 and #739.

**Comment by robertpfeiffer** [comment]
There are some modernisation issues with solarwolf unrelated to SDL2, like the news and the directories for screenshots and configuration. Do you want to update solarwolf to use more modern Python features and modules (without changing the gameplay or anything like that), or keep it in its historic state?

**Comment by illume** [comment]
Very nice work @dlon and @robertpfeiffer :)

**Comment by illume** [comment]
I did some more testing...

See the CREDS section is a different tint? I can see this only after losing a round and coming back the the main screen.
<img width="795" alt="screen shot 2019-01-21 at 07 46 50" src="https://user-images.githubusercontent.com/9541/51457296-26155700-1d51-11e9-9f8f-4c8f05a5f637.png">

The other menu colors are slightly off too (this is a SDL1 screenshot):
<img width="797" alt="screen shot 2019-01-21 at 07 45 04" src="https://user-images.githubusercontent.com/9541/51457416-9f14ae80-1d51-11e9-8567-977043fbbce2.png">


When loading there are some artifacts (might be to do with my 2013 Mac pro laptop).
<img width="758" alt="screen shot 2019-01-21 at 07 47 25" src="https://user-images.githubusercontent.com/9541/51457297-26155700-1d51-11e9-8809-93a65855d3aa.png">

**Comment by dlon** [comment]
> There are some modernisation issues with solarwolf unrelated to SDL2, like the news and the directories for screenshots and configuration. Do you want to update solarwolf to use more modern Python features and modules (without changing the gameplay or anything like that), or keep it in its historic state?

I was hoping illume would answer, but I'm guessing it wouldn't be a problem if it didn't break compatibility with old Python versions?

**Comment by dlon** [comment]
This works properly with SDL1 but not SDL2:

```python3
import pygame


pygame.display.init()
dsurf = pygame.display.set_mode((320, 240), 0)

def update():
    dsurf.fill((0, 0, 0))
    dsurf.blit(s, (0, 0))
    pygame.display.flip()

s = pygame.image.load('menu_creds_on.png')

s.set_colorkey(0, pygame.RLEACCEL)
#s.set_colorkey(0) # looks normal with this

s = s.convert()
update() # looks normal without this

s.set_alpha(10)
update()

s.set_alpha()
update()

pygame.event.clear()
pygame.event.wait()
```

**Comment by dlon** [comment]
There only seem to be issues using SDL_PIXELFORMAT_RGB888, if RLE is also used. Making surface.convert() use a different pixel format, like SDL_PIXELFORMAT_RGBX8888, causes the issues to disappear, but returns a different format than one would expect. Not using RLE also makes the issues to go away.

**Comment by MyreMylar** [comment]
Had a look at this and it looks like something in set_alpha() or set_colorkey() with the RLEACCEL flag is still broken. I'd debug it more but the game uses pngs and I still can't get windows to build from source with png support in SDL2.

**Simple Test case with solar wolf:**

1. Load up Solar Wolf.
2. enter the 'setup' menu then quit out.
3. enter the setup menu again then quit out - repeat this about ten times.
4. Your SolarWolf will now look something like this:

![image](https://user-images.githubusercontent.com/13382426/83426442-069f3b80-a427-11ea-8810-13b1cab9bd85.png)

Those colours the menu buttons change to are not random, they correspond to the color_keys of the image. So it appears that somehow when you set the RLEACCEL flag on a 32bit image with a color_key, and use set_alpha() the color_key value gets blended into the image like it was being treated like SRCALPHA...

hang on a minute:

https://github.com/pygame/pygame/issues/456

I think this is the same bug. SolarWolf is treating the colour key colour like the SRCALPHA value and blending it with the colours.

Simpler test case outside of solar wolf:

    import pygame
    
    
    pygame.display.init()
    dsurf = pygame.display.set_mode((320, 240), 0)
    
    
    def update(surface):
        dsurf.fill((0, 0, 0))
        dsurf.blit(surface, (0, 0))
        pygame.display.flip()
    
    
    img_surf = pygame.image.load('data/menu_creds_off.png')
    
    color_key = img_surf.get_colorkey()
    print('initial color_key', color_key)
    img_surf.set_colorkey(color_key, pygame.RLEACCEL)
    img_surf = img_surf.convert()
    print('initial alpha', img_surf.get_alpha())
    print('initial bitsize', img_surf.get_bitsize())
    print('initial flags', img_surf.get_flags())
    
    update(img_surf)
    
    img_surf.set_alpha(90)
    update(img_surf)
    img_surf.set_alpha(255)
    update(img_surf)
    
    running = True
    while running:
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                running = False

**Comment by illume** [comment]
The graphical glitches are fixed: 🎉

**Comment by illume** [comment]
I see two problems with solarwolf, so reopening:
- the bonuses that come out do not have transparent backgrounds... they have black backgrounds. It's hard to see because the background is mostly black.
- The 2.0.0.dev1 alpha blitter change introduced a bug that is visible with solarwolf.

**Comment by MyreMylar** [comment]
Looked into the first one.

Looks like in pygame 1.9.6 `surface.convert()` doesn't always strip per pixel alpha from a surface despite what the docs say. Solarwolf calls `surface.convert()` on the powerup effect images but it leaves the pixel alpha flag still on the surface which then causes the code to go down a different path.

Honestly, this feels like a bug in solarwolf more than in pygame, or perhaps `convert_alpha()` was added after solarwolf was created?

No idea what the other issue you mention is - couldn't see anything obvious.

**Comment by MyreMylar** [comment]
Added a pull request to solarwolf to fix the powerups so they work on 1.9.6 and 2.0.0:

https://github.com/pygame/solarwolf/pull/7

**Comment by illume** [comment]
(I updated my comment above with screenies)

1. Re: surface.convert() issue: 

I'd like to investigate this further to see under what circumstances this behaviour is happening. It looks like these are with 8bit colormap png files.
```bash
$ file solarwolf/data/debris1.png 
solarwolf/data/debris1.png: PNG image data, 80 x 10, 8-bit colormap, non-interlaced
$ file solarwolf/data/powerup.png 
solarwolf/data/powerup.png: PNG image data, 125 x 25, 8-bit colormap, non-interlaced
```

Maybe these images are supposed to be using color key transparency, rather than per pixel.

surf.convert() is supposed to convert to the fastest format for blitting, which can include alpha (now... not sure if it was a thing in SDL1).

ps. Using solarwolf here in this issue is meant to be one way to check backwards compatibility. Changing something in solarwolf doesn't fix all the other programs that may rely on this behaviour. Before merging that solarwolf PR, it would be good to understand the situation completely first.


2. re: select screen issue

note for self: I wonder if 7d69058 from #2269 is removed, will the other commits cause the new test in that PR to pass?

**Comment by MyreMylar** [comment]
> (I updated my comment above with screenies)

Hmm, this doesn't happen on windows. The select a player menu looks identical between dev1

I assume you are seeing this on mac? Had a look over the select player menu code code in solarwolf in windows and nothing obvious jumps out as wrong. It looks like two per pixel alpha text surfaces should get blitted to an opaque background surface and that's what happens on windows. I guess that the surfaces that come back from rendering text on mac and the default surfaces are different somehow which causes the oddness? The `textshadowed()` function would be the place to debug it on mac, see what the pixel colours of the blank space around the text - and surface formats all the way through are.

To answer your question about whether we can remove that commit I would say no. That one was the important commit to fix the issue with text seen in Flyboy and another game.

> Maybe these images are supposed to be using color key transparency, rather than per pixel.

The powerupeffects.png is 32 bit as well for some extra data. From what I stepped through the pngs are first loaded, then converted with `surface.convert()` which - per the docs - is supposed to strip any pixel alpha from the image. Then in solarwolf's `gfx.animstrip()` function it calls `get_alpha()`, which in pygame 1.9.6 returns 255 at this point, because the per pixel alpha hasn't been removed by `surface.convert()` as the docs say it is supposed to.

After that the image is divided up into sections and base images that have per pixel alpha are instead given instead a colorkey of (0,0,0) and stripped of any per pixel alpha. This part works fine in pygame 1 and pygame 2 if the incoming original image has per pixel alpha.

So from what I can tell it is just the different behaviour of surface.convert() between pygame 1 and 2. Pygame 2 seems correct to the docs. Relevant section here:

```
The converted Surface will have no pixel alphas. They will be stripped if the original had them. See convert_alpha() for preserving or creating per-pixel alphas.
```

But I'd welcome someone else looking into it because I may be missing something.

**Comment by illume** [comment]
Details of the set_mode surface:

```python
>>> screen
<Surface(320x200x32 SW)>
>>> screen.get_flags()
16842754
>>> screen.get_alpha()
255
>>> screen.get_shifts()
(16, 8, 0, 24)
>>> screen.get_losses()
(0, 0, 0, 0)
>>> screen.get_masks()
(16711680, 65280, 255, 4278190080)
>>> screen.get_blendmode()
1
```

**Comment by MyreMylar** [comment]
That seems like a bug, though I may be misunderstanding. Why does the
screen surface have alpha blending enabled? That's going to slow everything
down.

**Comment by MyreMylar** [comment]
I ran the same set of tests on the surface returned from `set_mode()` in solar wolf on windows:

```
screen: <Surface(800x600x32 SW)>
screen.get_flags(): 16777216
screen.get_flags() & pygame.SRCALPHA: 0
screen.get_alpha(): None
screen.get_shifts(): (16, 8, 0, 0)
screen.get_losses(): (0, 0, 0, 8)
screen.get_masks(): (16711680, 65280, 255, 0)
screen.get_blendmode(): 0
```

Assuming I understand this correctly it seems like `set_mode(`) is always incorrectly returning a surface with per pixel alpha on macs? I wonder if this is the reason for the corruption sometimes seen when people don't clear the screen surface on mac, some sort of garbage alpha data making a mess of things until the surface is filled with a colour.

**Comment by MyreMylar** [comment]
Looking at the SDL code the surface returned in `set_mode()` should always be an RGB surface, it ends up calling:

`SDL_CreateRGBSurfaceFrom()` so I'm unsure how we are ending up with a `set_mode()` surface with an alpha channel and blending on mac?

**Comment by MyreMylar** [comment]
Probably should add some asserts in to `set_mode()` during the refactor to check for this bug on all platforms - if it is confirmed.

**Comment by illume** [comment]
I'll look into it more a bit later, but it does seem weird. I mean, theoretically, I can see how a screen can be transparent. I mean... we have transparent windows, and actually transparent screens now... but it's probably a bug.

**Comment by MyreMylar** [comment]
Yeah, I was pondering if there was something I was missing with new tech.

I guess I could see some per pixel usage cases, like having some kind of mirror or see through hole in the centre with a fade to some GUI stuff around the edge or something. However, if we are going to support that (and it actually does work) I think we'd want to have it as an optional flag for the display surface rather than the default because blitting to an opaque surface is faster and the vast majority of users won't have a transparent/mirror screen.

**Comment by illume** [comment]
About the blend mode on for the screen on this Mac... 

I tried setting the screen surface blend mode to NONE at the end of set_mode and it doesn't seem to make a difference (to these two bugs).
```c
    SDL_SetSurfaceBlendMode(pgSurface_AsSurface(surface), SDL_BLENDMODE_NONE);
```

It does seem to turn alpha off on the screen surface though.
```python
>>> screen.get_alpha() == None
True
>>> screen.get_flags() & pygame.SRCALPHA
0
```
However, it chose the fastest format correctly. Blendmode on is the fastest format for this machine according to some quick benchmarks.

**Comment by illume** [comment]
1. Re: surface.convert() issue:

So... one difference is that get_alpha() is returning 255 on a convert()d 32bit surface with alpha on pygame 1.9.6 and 0 on 2.0, and also the flags are gone on 2.0, and are there with 1.9.6?

I confirmed it is so:
```python
>>> import pygame
pygame.init()
pygame 1.9.6
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> pygame.init()
(6, 0)
>>> s=pygame.display.set_mode((320,200))
>>> p=pygame.image.load('solarwolf/data/powereffects.png')
>>> cp = p.convert()
>>> cp.get_alpha()
255
>>> cp.get_flags() & pygame.SRCALPHA
65536
>>> p.get_alpha()
255
>>> cp.get_alpha()
255
>>> p.get_flags() & pygame.SRCALPHA
65536
```

Here with pygame `main`...
```python
>>> import pygame
pygame 2.0.1.dev1 (SDL 2.0.12, python 3.8.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> pygame.init();pygame.display.set_mode((320,200))
(6, 0)
2020-11-17 18:51:41.299 Python[33505:1383705] ApplePersistenceIgnoreState: Existing state will not be touched. New state will be written to (null)
<Surface(320x200x32 SW)>
>>> p=pygame.image.load('solarwolf/data/powereffects.png')
>>> cp = p.convert()
>>> cp.get_alpha()
>>> cp.get_flags() & pygame.SRCALPHA
0
>>> p.get_flags() & pygame.SRCALPHA
65536
>>> p.get_alpha()
255
```

I'm currently thinking this is behaviour that we should try and keep. The docs aren't clear what 'strip alpha' means exactly. What it does do in 1.9.6 is not use the alpha when drawing, but keep the alpha info in the surface flags. So keeping this (if possible), allows keeping backwards compatibility and would still hold with the interpretation of the docs that alpha is stripped in the way it doesn't draw with per pixel alpha transparency.

**Comment by MyreMylar** [comment]
How does it flag these alpha surfaces that shouldn't be drawn with alpha? Is there a hidden 'alpha_but_not_actually' flag in pygame 1?

**Comment by MyreMylar** [comment]
> However, it chose the fastest format correctly. Blendmode on is the fastest format for this machine according to some quick benchmarks.

This seems counter intuitive to me because it would seem likely to trigger making blits with an alpha channel on the source as well run down a slower blending path - but it's hard to know what is going on at that level as I don't own a mac and windows appears to work differently. Maybe the `display.flip()/display,update()` is faster with an alpha blendmode enabled for some reason on mac.

**Comment by illume** [comment]
> This seems counter intuitive to me because it would seem likely to trigger making blits with an alpha channel on the source as well run down a slower blending path - but it's hard to know what is going on at that level as I don't own a mac and windows appears to work differently. Maybe the display.flip()/display,update() is faster with an alpha blendmode enabled for some reason on mac.

To be fair, I didn't do extensive benchmarking. Yeah, there are often fast paths in most graphics drivers with regards to formats. For the longest time ARGB was very fast on most Mac drivers (but not all). It's the same on other platforms too. I'm pretty sure Metal does some sort of JIT internally, where the speed increases massively after some warmup period of about 7 seconds.

**Comment by illume** [comment]
> How does it flag these alpha surfaces that shouldn't be drawn with alpha? Is there a hidden 'alpha_but_not_actually' flag in pygame 1?

I'm not sure exactly, but likely something to do with the `->format`.  I'll first make a test case and then try and debug it.

TODO list...
- [ ] make a test for get_flags and get_alpha (opaque (alpha=255) surfaces return 255 not None) with a convert()'d png style surface.
- [ ] debug the `textshadowed()` to make a test out of the menu display issue.
- [ ] more benchmarking without the blend_mode on mac.

**Comment by MyreMylar** [comment]
Had an idea to rejig the if conditionals in alphablend.c that might help
with the shadowed text on Mac issue. I think I recall you saying that it
works in 2.0.0 but not in 2.0.0.dev1.

Less sure what we should do about convert though as a lot of our blits
still go through SDL2. If we actually mess with the surface formats and
flags used by SDL2 it's likely to cause chaos and slow down a lot of blits.

Maybe a solarwolf_once_had_alpha_hack flag is the way forward that is set
on surfaces in convert() and checked in get_alpha()?

**Comment by illume** [comment]
Note here on SDL1, get_losses() returns 8 for the `surf->format->Aloss` after being convert()d.

```python
>>> import pygame
pygame 2.0.1.dev1 (SDL 1.2.15, python 3.8.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> pygame.display.init()
>>> pygame.display.set_mode((640, 480))
<Surface(640x480x32 SW)>
>>> s1 = pygame.Surface((100, 100), pygame.SRCALPHA, 32)
>>> s1.get_flags()
65536
>>> s1.get_masks()
(16711680, 65280, 255, 4278190080)
>>> s1.get_losses()
(0, 0, 0, 0)
>>> s1.get_shifts()
(16, 8, 0, 24)
>>> sc = s1.convert()
>>> sc.get_losses()
(0, 0, 0, 8)
>>> sc.get_shifts()
(8, 16, 24, 0)
>>> 
```

**Comment by illume** [comment]
Here's the SDL2 version. Note that get_losses() has 0 for `surf->format->Aloss`.
```python
>>> import pygame
pygame 2.0.1.dev1 (SDL 2.0.12, python 3.8.6)
Hello from the pygame community. https://www.pygame.org/contribute.html
>>> pygame.display.set_mode((640, 480))
<Surface(640x480x32 SW)>
>>> s1 = pygame.Surface((100, 100), pygame.SRCALPHA, 32)
>>> s1.get_losses()
(0, 0, 0, 0)
>>> sc = s1.convert()
>>> sc.get_losses()
(0, 0, 0, 0)
```

It seems the secret is `surf->format->Aloss`.

**Comment by illume** [comment]
Here are the SDL1, pygame 1.9.6 screen losses on this mac.
```
>>> screen = pygame.display.set_mode((320,200))
>>> screen.get_losses()
(0, 0, 0, 8)
```


It seems solarwolf relied on convert() working with alpha and non-alpha surfaces.

From https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdldisplayformat.html
> This function takes a surface and copies it to a new surface of the pixel format and colors of the video framebuffer, suitable for fast blitting onto the display surface. It calls SDL_ConvertSurface
> 
> If you want to take advantage of hardware colorkey or alpha blit acceleration, you should set the colorkey and alpha value before calling this function.
> 
> If you want an alpha channel, see SDL_DisplayFormatAlpha.

The line "set the colorkey and alpha value before calling this function" may apply to what SDL_image does. Can a png have a color key with "8-bit/color RGBA" via SDL_image?

From SDL2_image-2.0.5/IMG_png.c
>    /* For images with a single "transparent colour", set colour key;
>       if more than one index has transparency, or if partially transparent
>      entries exist, use full alpha channel */

solarwolf/data/powereffects.png is an image with 1 "transparent color", so the color key is likely being set.
Except with pygame 1.9.6 it doesn't set a color key. It does set an Amask however.

https://wiki.libsdl.org/SDL_CreateRGBSurface
> By default surfaces with an alpha mask are set up for blending as with
>    SDL_SetSurfaceBlendMode(surface, SDL_BLENDMODE_BLEND) 

Confirmed in SDL2-2.0.12/src/video/SDL_surface.c@SDL_ConvertSurface seems to keep alpha, and a blend mode.

It seems that convert() should definitely keep alpha transparency, and a blend mode when it's there.

**Comment by illume** [comment]
> Had an idea to rejig the if conditionals in alphablend.c that might help
with the shadowed text on Mac issue. I think I recall you saying that it
works in 2.0.0 but not in 2.0.0.dev1.

Yeah, it works in 2.0.

**Comment by illume** [comment]
Note the patch in #2345 by @MyreMylar seems to fix it on my Mac display.

Did some testing with `textshadowed()`. The main difference seems to be the get_at() on current `main` fails to give an alpha, whereas the patch from #2345 and 1.9.6 do. src_surf

Working up a PR with a test now...

With patch:
```python
Select A Player:
> solarwolf/solarwolf/txt.py(124)textshadowed()
-> return [img, r]
(Pdb) img
<Surface(268x37x32 SW)>
(Pdb) img.get_flags() & pygame.RLEACCELOK
8192
(Pdb) img.get_flags() & pygame.SRCCOLORKEY
4096
(Pdb) img1.get_flags() & pygame.SRCALPHA
65536
(Pdb) img2.get_flags() & pygame.SRCALPHA
65536

(Pdb) img.get_flags()
12288
(Pdb) img.get_losses()
(0, 0, 0, 0)
(Pdb) img.get_masks()
(16711680, 65280, 255, 4278190080)
(Pdb) img.get_flags() & pygame.SRCALPHA
0
(Pdb) img.get_colorkey()
(0, 0, 0, 255)
(Pdb) img.get_at((10,10))
(0, 0, 0, 0)
(Pdb) img.get_at((5,5))
(220, 210, 180, 0)

(Pdb) img1, img2
(<Surface(266x35x32 SW)>, <Surface(266x35x32 SW)>)

(Pdb) img2.get_flags(), img2.get_losses(), img2.get_masks(), img2.get_at((5,5)), img2.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (110, 105, 90, 255), None)

(Pdb) img1.get_flags(), img1.get_losses(), img1.get_masks(), img1.get_at((5,5)), img1.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (220, 210, 180, 255), None)
```

Without patch:
```python
Select A Player:
> solarwolf/solarwolf/txt.py(124)textshadowed()
-> return [img, r]
(Pdb) img
<Surface(268x37x32 SW)>
(Pdb) img.get_losses()
(0, 0, 0, 0)
(Pdb) img.get_masks()
(16711680, 65280, 255, 4278190080)
(Pdb) img.get_colorkey()
(0, 0, 0, 255)
(Pdb) img.get_at((5,5))
(220, 210, 180, 255)
(Pdb) img.get_flags() & pygame.SRCALPHA
0
(Pdb) img.get_flags()
12288

(Pdb) img1, img2
(<Surface(266x35x32 SW)>, <Surface(266x35x32 SW)>)

(Pdb) img2.get_flags(), img2.get_losses(), img2.get_masks(), img2.get_at((5,5)), img2.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (110, 105, 90, 255), None)

(Pdb) img1.get_flags(), img1.get_losses(), img1.get_masks(), img1.get_at((5,5)), img1.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (220, 210, 180, 255), None)

```

1.9.6 mac python 3.8
```python
(Pdb) img2.get_flags(), img2.get_losses(), img2.get_masks(), img2.get_at((5,5)), img2.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (110, 105, 90, 255), None)
(Pdb) img1.get_flags(), img1.get_losses(), img1.get_masks(), img1.get_at((5,5)), img1.get_colorkey()
(65536, (0, 0, 0, 0), (16711680, 65280, 255, 4278190080), (220, 210, 180, 255), None)

(Pdb) img.get_flags(), img.get_losses(), img.get_masks(), img.get_at((5,5)), img.get_colorkey()
(12288, (0, 0, 0, 8), (65280, 16711680, 4278190080, 0), (220, 210, 180, 255), (0, 0, 0, 255))
```

**Comment by illume** [comment]
TLDR; I think there's a problem with `pg_DisplayFormat`, which is leaving some alpha parts in the pixel format.

---

Did lots of tests, and found out some things.

Basically, the convert() is leaving the losses AND the alpha part of the mask wrong.

They are coming out like:
```python
masks: (16711680, 65280, 255, 4278190080)
losses: (0, 0, 0, 0)
```

But, they should be something like this:
```python
            self.assertEqual(img.get_masks(), (65280, 16711680, 4278190080, 0))
            self.assertEqual(img.get_losses(), (0, 0, 0, 8))
```

Note, the 8 in the losses will cancel out alpha.

```python
    def test_src_alpha_rleaccel(self):
        """
        """
        pygame.display.init()
        pygame.font.init()
        try:
            screen = pygame.display.set_mode((640, 480), flags=pygame.SRCALPHA, depth=32)
            # screen.set_losses((0, 0, 0, 8))
            # screen.set_blendmode()
            expected_color = (80, 77, 66, 255)
            color = (220, 210, 180)
            font = pygame.font.Font(None, 20)
            img1 = font.render('Select A Player:', 1, color)

            # img1 = pygame.Surface((20, 20), flags=pygame.SRCALPHA, depth=32)
            # img1.set_at((10, 10), expected_color)

            # img = pygame.Surface((20, 20), flags=0, depth=32)
            img = pygame.Surface((20, 20))
            # img = pygame.Surface((20, 20), flags=pygame.SRCALPHA, depth=32, masks=(4278190080, 16711680, 65280, 0))
            # img = pygame.Surface((20, 20), flags=pygame.SRCALPHA, depth=32)
            # if hasattr(img, 'set_losses'):
            #     img.set_losses((0, 0, 0, 0))
            # else:
            #     self.assertEqual(img.get_losses(), (0, 0, 0, 8))

            img.blit(img1, (0, 0))
            img = img.convert()
            # img.set_losses((0, 0, 0, 8))
            # self.assertEqual(img.get_masks(), (16711680, 65280, 255, 4278190080))
            self.assertEqual(img.get_losses(), (0, 0, 0, 8))
            # self.assertTrue(img.get_flags() & pygame.SRCALPHA)

            img.set_colorkey((0, 0, 0), pygame.RLEACCEL)
            self.assertEqual(img.get_at((10, 10)), expected_color)
        finally:
            pygame.font.quit()
            pygame.display.quit()
```

**Comment by zoldalma999** [comment]
> Perhaps we should just leave the solarwolf issue open so we can figure out what the test should be and then merge this for now anyway.

Shouldn't this be reopened?

**Comment by illume** [comment]
Sorry, I didn't notice the other issue was linked to close this one.
